File: hooks/block-git-commit-protected-changes.py
#!/usr/bin/env python3
"""PreToolUse hook: block git commit when staged diff contains protected config changes."""

import json
import re
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from nix_protected import check_config_diff, check_mkforce_override

NIX_CONFIG = Path.home() / "nix-config"


def get_staged_diff(cwd: Path) -> str:
    result = subprocess.run(
        ["git", "diff", "--cached"],
        capture_output=True, text=True, cwd=cwd,
    )
    return result.stdout if result.returncode == 0 else ""


def main() -> None:
    data = json.load(sys.stdin)
    command = data.get("tool_input", {}).get("command", "")

    if not re.search(r"\bgit\s+commit\b", command):
        return

    diff_text = get_staged_diff(NIX_CONFIG)
    if not diff_text.strip():
        return

    reason = check_config_diff(diff_text)
    if not reason:
        reason = check_mkforce_override(diff_text)

    if reason:
        json.dump({"decision": "block", "reason": reason}, sys.stdout)


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/block_git_commit_protected_changes.rs
use regex::Regex;

use crate::input::HookInput;
use crate::nix_protected::{check_config_diff, check_mkforce_override};
use crate::output::{block, pass};

pub fn run(input: &HookInput) {
    let command = &input.tool_input.command;
    if !Regex::new(r"\bgit\s+commit\b").unwrap().is_match(command) {
        return;
    }

    let nix_config = home::home_dir().unwrap_or_default().join("nix-config");
    let output = std::process::Command::new("git")
        .args(["diff", "--cached"])
        .current_dir(&nix_config)
        .output();

    let diff_text = match output {
        Ok(o) if o.status.success() => String::from_utf8_lossy(&o.stdout).to_string(),
        _ => return,
    };

    if diff_text.trim().is_empty() {
        return;
    }

    if let Some(reason) = check_config_diff(&diff_text) {
        block(&reason);
        return;
    }
    if let Some(reason) = check_mkforce_override(&diff_text) {
        block(&reason);
        return;
    }
    pass();
}
